### ISSUE TITLE
Cosmos is confused by Higher-Order Components

### ISSUE THREAD

**Issue by alp82** [react-cosmos/react-cosmos#494]
### What's up?

At least, that's what i think happens. I have fixtures of two components that are wrapped with the same HOC. Those are not appearing in the redux-cosmos page, at least not as they should be. One set is missing completly and the other one is labeled wrongly. It uses the name of the HOC instead of the component it includes.

I am using the current version (beta5).

### Mkay, tell me more...

This is the folder structure. Note that my `__fixtures__` are named `__cosmox__`:
![image](https://user-images.githubusercontent.com/966342/32415046-17d8a220-c233-11e7-9bb1-7b672f52aa62.png)

The result in cosmos:
![image](https://user-images.githubusercontent.com/966342/32415117-9739809c-c234-11e7-97d0-a755950edacf.png)

These are appearently the fixtures of `Textarea`, but labeled wrongly. The fixtures of `Input` are missing completely. They are correct though, specifying the `component` they belong to.

`Input` and `Textarea` look very similar (simplified):
```
import glamorous from 'glamorous';
import FloatingLabelHOC from 'ui/form/FloatingLabelHOC';

const Field = glamorous.textarea(({ theme, padding }) => {
  return {
    paddingTop: value ? `calc(${padding} * (1 + 3/4))` : padding,
    paddingBottom: value ? `calc(${padding} * (1 - 3/4))` : padding,
    transition: `padding ${theme.animation.duration.regular} ${theme.animation.easing.regular}`,
  };
}));

const Textarea = FloatingLabelHOC(Field);
export default Textarea;
```

And FloatingLabelHOC (simplified):
```
const FloatingLabelHOC = Field => {
  class FloatingLabel extends Component {
    static propTypes = {
      name: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      value: PropTypes.string,
    };

    static defaultProps = {
      value: '',
    };

    componentWillMount() {
      this.setState({
        value: get(this, 'props.value', ''),
      });
    }

    render() {
      const { name, label } = this.props;
      const { value } = this.state;

      return (
        <div>
          <Field
            {...{
              name,
              value,
              label,
            }}
          />
          {label && (
            <Label value={ value}>
              <LabelText>{label}</LabelText>
            </Label>
          )}
        </div>
      );
    }
  }
  return FloatingLabel;
};
```

I assumed that it might be related to setting `displayName` manually and voilà, that works:
```
Textarea.displayName = 'Textarea';
```
(and the same for `Input`)

Is this intended to work only this way? I don't like writing additional code only for the purpose of fixing my tests.

**Comment by skidding** [comment]
I think I found the fix for this. It's twopart:

1. Made sure components with same name never overlap again (#497). Please install `react-cosmos-webpack@3.0.0-beta.7` and see if all your fixtures show up **without** setting a custom displayName on each
2. HoCs have the responsibility to extend a component's name. [Here is an example from React Router](https://github.com/ReactTraining/react-router/blob/5470fad8876d7b788e29c4a61e83fa62a4301504/packages/react-router/modules/withRouter.js#L19). My suggestion is to do this:

```diff
const FloatingLabelHOC = Field => {
  class FloatingLabel extends Component {
    static propTypes = {
      name: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      value: PropTypes.string,
    };

    static defaultProps = {
      value: '',
    };

    componentWillMount() {
      this.setState({
        value: get(this, 'props.value', ''),
      });
    }

    render() {
      const { name, label } = this.props;
      const { value } = this.state;

      return (
        <div>
          <Field
            {...{
              name,
              value,
              label,
            }}
          />
          {label && (
            <Label value={ value}>
              <LabelText>{label}</LabelText>
            </Label>
          )}
        </div>
      );
    }
  }
+ FloatingLabel.displayName = `FloatingLabel(${Field.displayName || Field.name})`
  return FloatingLabel;
};
```

... and you'll never have to set `displayName` by hand after that.

Does it work?

**Comment by alp82** [comment]
1. Yep, updating to beta7 solves the issue with overlapping names. I got two entries with a counter as suffix and all fixtures are selectable.

2. Kinda, but not really:
![image](https://user-images.githubusercontent.com/966342/32417885-2f6d0de0-c260-11e7-9c96-f516abbd7487.png)

**Comment by skidding** [comment]
> Kinda, but not really:

I might be subjective but this seems correct. Ugly, but correct. It tells you that the Input is wrapped in glamorous and FloatingLabel HoCs, which helps you understand its API (props, state, etc) better.

I see a number of possible avenues from here:
- Use the `HoC2(HoC1(Component))` convention to render prettier component names. Component name could be big while HoC names on a smaller second line, or shown when hovering in a tooltip. But, the new version needs more mileage before investing into something so particular.
- You can always override the name via `displayName`.
- It's annoying to rename all components if you use many HoCs and don't like how the composed names look. There is another possibility. From [this comment](https://github.com/react-cosmos/react-cosmos/issues/495#issuecomment-341994921) you can see we already gather component names statically. Those names are simpler, they are the names we give component imports in our fixtures. We could add a Cosmos config to rely on those instead of run time names. Eg. `preferRuntimeComponentNames` which I'd still keep `true` by default but you might prefer to set to `false`. Note: Component names gathered statically are even more experimental and can be confusing. If you do `import MyComponent from './ActualComponentName.jsx'` it will show up as `MyComponent` in the UI.

Wdyt?

**Comment by alp82** [comment]
Yes this is reasonable.

I would love to see the `preferRuntimeComponentNames` option implemented!

**Comment by skidding** [comment]
> I would love to see the `preferRuntimeComponentNames` option implemented!

If it's not too much to ask, can you please open a new issue for this feature request specifically and close this one? Feel free to reference this issue and use our conversation is quotes in the description. Thanks!

**Comment by alp82** [comment]
Overlapping issues are solved, further discussion in #508.
